programs/fuzz/Makefile

```diff
@@ -1,40 +1,17 @@
-MBEDTLS_TEST_PATH:=../../tests/src
-MBEDTLS_TEST_OBJS:=$(patsubst %.c,%.o,$(wildcard ${MBEDTLS_TEST_PATH}/*.c ${MBEDTLS_TEST_PATH}/drivers/*.c))
+MBEDTLS_TEST_PATH:=../../tests
 
-CFLAGS ?= -O2
-WARNING_CFLAGS ?= -Wall -Wextra
 LOCAL_CFLAGS = $(WARNING_CFLAGS) -I../../tests/include -I../../include -D_FILE_OFFSET_BITS=64
-LOCAL_LDFLAGS = ${MBEDTLS_TEST_OBJS}		\
-		-L../../library			\
-		-lmbedtls$(SHARED_SUFFIX)	\
-		-lmbedx509$(SHARED_SUFFIX)	\
-		-lmbedcrypto$(SHARED_SUFFIX)
+MBEDTLS_PATH := ../..
+include ../../scripts/common.make
 
 LOCAL_CFLAGS += $(patsubst -I../%,-I../../%,$(THIRDPARTY_INCLUDES))
```

Comment:
Maybe we could move this to `common.make` and do the substitution with the help of MBEDTLS_PATH.

Reply:
Doesn't the line in `common.make` already work?

Reply:
Indeed it should already work. (It wasn't immediately obvious to me.) In that case we don't need this line either.

Reply:
Ok, so apparently it doesn't work after all. The parent Makefile already included `common.make` and all variables are set. We use `?=` in `common.make` so when we do the include here, we still get the parent values.

We either need to
1. do the substitution (in which case the include might be superfluous) or
2. change `common.make` to override values (we can always start with the include if we want to customise the variables)

@gilles-peskine-arm , @paul-elliott-arm what do you think?

Reply:
The problem isn't how `LOCAL_CFLAGS` is calculated, it's that `THIRDPARTY_INCLUDES` is already set to a different value that is wrong. This fixes it when building the whole tree (`make programs`):

```
diff --git a/programs/Makefile b/programs/Makefile
index 82c8569963..f6eb2f4dad 100644
--- a/programs/Makefile
+++ b/programs/Makefile
@@ -114,7 +114,7 @@ all: fuzz
 endif
 
 fuzz: ${MBEDTLS_TEST_OBJS}
-       $(MAKE) -C fuzz THIRDPARTY_INCLUDES=$(THIRDPARTY_INCLUDES)
+       $(MAKE) -C fuzz THIRDPARTY_INCLUDES=$(THIRDPARTY_INCLUDES:-I../%=-I../../%)
 
 ${MBEDTLS_TEST_OBJS}:
        $(MAKE) -C ../tests mbedtls_test
```

I think this gets the CI to pass, but running `make` directly inside `programs/fuzz` is still broken. For that, I think we should fix the definition of `THIRDPARTY_INCLUDES`, which has `../` built in. I don't understand why it doesn't use `$(THIRDPARTY_DIR)`. So I think the right fix is to make the definition of `THIRDPARTY_INCLUDES` use `$(THIRDPARTY_DIR)` and drop the hack inside `programs/Makefile`.

```
diff --git a/3rdparty/everest/Makefile.inc b/3rdparty/everest/Makefile.inc
index 77a6b49654..8055ce9503 100644
--- a/3rdparty/everest/Makefile.inc
+++ b/3rdparty/everest/Makefile.inc
@@ -1,6 +1,6 @@
-THIRDPARTY_INCLUDES+=-I../3rdparty/everest/include -I../3rdparty/everest/include/everest -I../3rdparty/everest/include/everest/kremlib
+THIRDPARTY_INCLUDES+=-I$(THIRDPARTY_DIR)/everest/include -I$(THIRDPARTY_DIR)/everest/include/everest -I$(THIRDPARTY_DIR)/everest/include/everest/kremlib
 
 THIRDPARTY_CRYPTO_OBJECTS+= \
-       ../3rdparty/everest/library/everest.o \
-       ../3rdparty/everest/library/x25519.o \
-       ../3rdparty/everest/library/Hacl_Curve25519_joined.o
+       $(THIRDPARTY_DIR)/everest/library/everest.o \
+       $(THIRDPARTY_DIR)/everest/library/x25519.o \
+       $(THIRDPARTY_DIR)/everest/library/Hacl_Curve25519_joined.o
diff --git a/3rdparty/p256-m/Makefile.inc b/3rdparty/p256-m/Makefile.inc
index fc8f73bf84..53bb55b547 100644
--- a/3rdparty/p256-m/Makefile.inc
+++ b/3rdparty/p256-m/Makefile.inc
@@ -1,5 +1,5 @@
-THIRDPARTY_INCLUDES+=-I../3rdparty/p256-m/p256-m/include -I../3rdparty/p256-m/p256-m/include/p256-m -I../3rdparty/p256-m/p256-m_driver_interface
+THIRDPARTY_INCLUDES+=-I$(THIRDPARTY_DIR)/p256-m/p256-m/include -I$(THIRDPARTY_DIR)/p256-m/p256-m/include/p256-m -I$(THIRDPARTY_DIR)/p256-m/p256-m_driver_interface
 
 THIRDPARTY_CRYPTO_OBJECTS+= \
-       ../3rdparty/p256-m//p256-m_driver_entrypoints.o \
-       ../3rdparty/p256-m//p256-m/p256-m.o
+       $(THIRDPARTY_DIR)/p256-m//p256-m_driver_entrypoints.o \
+       $(THIRDPARTY_DIR)/p256-m//p256-m/p256-m.o
diff --git a/programs/Makefile b/programs/Makefile
index 82c8569963..6baf4651a5 100644
--- a/programs/Makefile
+++ b/programs/Makefile
@@ -114,7 +114,7 @@ all: fuzz
 endif
 
 fuzz: ${MBEDTLS_TEST_OBJS}
-       $(MAKE) -C fuzz THIRDPARTY_INCLUDES=$(THIRDPARTY_INCLUDES)
+       $(MAKE) -C fuzz
 
 ${MBEDTLS_TEST_OBJS}:
        $(MAKE) -C ../tests mbedtls_test
```

This makes `make lib programs && make -C programs/fuzz clean && make -C programs/fuzz` work on this PR with Everest enabled. I can't think of anything that currently works and that this patch breaks, but I haven't tested significantly.

Reply:
> running make directly inside programs/fuzz is still broken

Note that this is a thing oss-fuzz does, so we mustn't break it. And we don't test that in our CI.